### PR TITLE
Movielens.py to load movie title and genres columns

### PR DESCRIPTION
**Pull request by loomlike**
### Description
Movielens.py to load movie title and genres columns when title_col and/or genres_col are given

### Motivation and Context
Movie title can be useful for debugging, while genres can be used as item features for certain models.

### Releated Issues
#403

### Checklist:

- [x] My code follows the code style of this project, as detailed in our [contribution guidelines](../CONTRIBUTING).
- [x] I have added tests.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: reco_utils/dataset/movielens.py [hunk 1]
```diff
@@ -18,18 +18,41 @@
...
 
 
 # MovieLens data have a different format for each size
 class DataFormat:
-    def __init__(self, sep, path, has_header=False):
+    def __init__(
+            self,
+            sep, path, has_header=False,
+            item_sep=None, item_path=None, item_has_header=False,
+            # user_sep=None, user_path=None, user_has_header=False
```
Comment (yueguoguo) [inline reco_utils/dataset/movielens.py]: Remove commented codes.

File: reco_utils/dataset/movielens.py [hunk 2]
```diff
@@ -18,18 +18,41 @@
...
         self._sep = sep
         self._path = path
         self._has_header = has_header
 
+        # Item file
+        self._item_sep = item_sep
+        self._item_path = item_path
+        self._item_has_header = item_has_header
+
+        # User file
```
Comment (yueguoguo) [inline reco_utils/dataset/movielens.py]: Remove commented codes (if not needed).

File: reco_utils/dataset/movielens.py [hunk 3]
```diff
@@ -42,73 +65,189 @@ def path(self):
...
+
+    @property
+    def item_path(self):
+        return self._item_path
 
+    @property
+    def item_has_header(self):
+        return self._item_has_header
+
+    # """ User file """
```
Comment (yueguoguo) [inline reco_utils/dataset/movielens.py]: Remove commented codes.

File: reco_utils/dataset/movielens.py [hunk 4]
```diff
@@ -18,18 +18,41 @@
...
+        col,
+    )
+except:
     pass  # so the environment without spark doesn't break
 
 
 # MovieLens data have a different format for each size
 class DataFormat:
-    def __init__(self, sep, path, has_header=False):
+    def __init__(
```
Comment (yueguoguo) [inline reco_utils/dataset/movielens.py]: docstrings

**Comment by yueguoguo** [comment]
Awesome @loomlike only a few minor coding style comments are provided.

**Comment by miguelgfierro** [comment]
Adding @danielsc for review

File: reco_utils/dataset/movielens.py [hunk 5]
```diff
@@ -18,18 +18,46 @@
         FloatType,
         DoubleType,
         LongType,
+        StringType,
     )
-except ModuleNotFoundError:
```
Comment (miguelgfierro) [inline reco_utils/dataset/movielens.py]: Can you add back ModuleNotFoundError? this should pass when we are in the bare env but if there is another exception, we should know

File: reco_utils/dataset/movielens.py [hunk 6]
```diff
@@ -42,73 +70,174 @@ def path(self):
...
+    def item_path(self):
+        return self._item_path
+
+    @property
+    def item_has_header(self):
+        return self._item_has_header
+
 
+# 10m and 20m data do not have user data
 _data_format = {
```
Comment (miguelgfierro) [inline reco_utils/dataset/movielens.py]: if what you want is a constant dict, `_data_format` should be `DATA_FORMAT`, same with `_genres`. Is that what you want?

File: reco_utils/dataset/movielens.py [hunk 7]
```diff
@@ -42,73 +70,174 @@ def path(self):
...
-ERROR_MOVIE_ID_TYPE = "Movie id should be IntegerType"
-ERROR_RATING_TYPE = "Rating should be FloatType or DoubleType"
+ERROR_LOCAL_CACHE_PATH = """Local cache path only accepts a zip file path:
+    use/something/like_this.zip"""
 
 
-def load_pandas_df(size="100k", header=None, local_cache_path="ml.zip"):
+def load_pandas_df(
+    size="100k",
+    header=("UserId", "MovieId", "Rating", "Timestamp"),
```
Comment (miguelgfierro) [inline reco_utils/dataset/movielens.py]: Question here: is there any relationship of these names `"UserId", "MovieId", "Rating", "Timestamp"` to those in this file https://github.com/Microsoft/Recommenders/blob/00bf677a714b7d1817c1e14115228b6724f0ad45/reco_utils/common/constants.py?

File: tests/unit/test_movielens.py [hunk 8]
```diff
@@ -13,135 +13,121 @@
         FloatType,
         DoubleType,
     )
+    from pyspark.sql.functions import col
     from reco_utils.common.spark_utils import start_or_get_spark
 except:
```
Comment (miguelgfierro) [inline tests/unit/test_movielens.py]: same here, we should add ModuleNotFoundError

**Comment by miguelgfierro** [comment]
@loomlike Jun, after this is merged we should update the notebooks to add the feature you created

**Comment by loomlike** [comment]
@miguelgfierro @yueguoguo Thanks for the review both of you. And yeah we may need to update some of notebooks, especially those who uses default col names because the movievlens loader now uses DEFAULT_XXX_COL variables defined in Constant.py, which starts with the lower case. Will see if notebook tests are passed and if not I will update them accordingly.

**Comment by miguelgfierro** [comment]
there is an issue with the tests: `IllegalArgumentException: 'Field "UserId" does not exist.\nAvailable fields: userID, itemID, rating, timestamp'`

**Comment by loomlike** [comment]
@miguelgfierro Updated notebook fix

**Pull request merged**